Allow configuration of resolution for existing data in field mapping

Alexalex1432 opened the pull request:
This will let users preserve existing data if they choose, or overwrite if the data from Funraise is more recent or not empty.
Resolving FUN-12869

Commit:
Allow configuration of resolution for existing data in field mapping

This will let users preserve existing data if they choose, or overwrite if the data from Funraise is more recent or not empty.
Resolving FUN-12869

Alexalex1432 commented:
Beta package install link will be here shortly, working on arranging the order of other improvements that we may want to release before this piece of functionality

elanmajkrzak commented on src/classes/frModel.cls:
```diff
@@ -19,10 +24,21 @@ public with sharing abstract class frModel {
     }
     
+    protected virtual Datetime getUpdtime() {
+        if(getRequestBody().containsKey('updtime')) {
```
Since it looks like we're essentially just going to prefix this for every implementing model, outside of donation, maybe it's best to have an abstract method that returns the key, so we don't have to implement the Datetime building in every model

elanmajkrzak commented on src/classes/frDonorTest.cls:
```diff
@@ -239,8 +239,127 @@ public class frDonorTest {
+        Map<String, Object> request = getTestRequest();
+        request.put('id', existing.fr_Id__c);
+        request.put('email', newEmail); //should overwrite existing value
+        request.put('updtime', DateTime.now().addDays(1).getTime()); //one day in the future, so it should always be more recent
```
I seriously doubt we'd ever run into this, but it's probably best to take the time you this to be later than and add to it, rather than using now.
`existing.addDays(1).getTime()`

Alexalex1432 replied:
Ok this was the only change I didn't implement from the feedback.  It was going to require us to requery the record from the database because that's how LastModifiedDate gets populated, but we know it was last modified at the time the test is running (aka DateTime.now()) since it gets inserted a few lines above this call so I think we should be fine.  If you feel strongly, I can definitely implement the requery-after-insert

elanmajkrzak replied:
Oh that's fine, I thought it was coming from the `getTestContact` method. Pretty sure you'll never hit it anyways since that's a static value as is

adelorme82 commented on src/classes/frModel.cls:
```diff
@@ -56,32 +72,84 @@ public with sharing abstract class frModel {
+        String salesforceId = getSalesforceId();
+        SObject queriedFieldsRecord = null;
+        if(!fieldsToQueryFor.isEmpty()) {
+            String query = 'SELECT ' + String.join(new List<String>(fieldsToQueryFor), ',') 
```
I'm not super familiar with the syncs, do we ever sync more than one? Do we need to worry about bulkifying this?

Alexalex1432 replied:
Currently, no.  It's one mappable record per sync/lifecycle.   And I think even if we did I would imagine this would mostly stay the same, but we would update the query to be `WHERE Id IN (list of things we're syncing)` since the columns in this query are pretty static for every incoming sync based on how they setup their mappings

adelorme82 commented on src/classes/frSetupController.cls:
```diff
@@ -36,6 +36,31 @@ public with sharing class frSetupController {
         private set;
     }
+    
+    public static String MAPPING_OVERWRITE {
```
Super minor, muh whitespaces

adelorme82 commented on src/classes/frModel.cls:
```diff
@@ -56,32 +72,84 @@ public with sharing abstract class frModel {
+                        write = existingValue == null;
+                    } else if (mapping.Conflict_Resolution__c == MAPPING_OVERWRITE_RECENT && isMoreRecent) {
+                        write = true;
+                    }
```
Should we add logging/exceptions/something for if the conflict resolution value is none of those?

Alexalex1432 replied:
Ha, this was actually a cause of a unit test failure that took me a while to unwind.  I had specified the default value for the Conflict_Resolution__c field and it didn't match the defined values in code and never hit the correct branch in here.  I think adding a log is a good idea

adelorme82 commented on src/pages/frSetup.page:
```diff
@@ -48,6 +48,17 @@
 						<apex:selectOptions value="{!donationSFOptions}" />
 					</apex:selectList> 
 				</apex:column>
+				<apex:column >
```
more whitespace here unfortunately

Commit:
PR Feedback:

- Change the way we retrieve updtime to only needing derived classes to provide their json property key
- Fix whitespace in a few classes + setup page
- Add logging if we encounter a conflict resolution that we don't recognize
Resolving FUN-12869

Alexalex1432 commented:
Beta Install Link: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t6R000000Y2jl
